# Development of new data processing scripts for climate stations at GVC


This notebook is just used for testing during the development of new processing scripts. 

In [1]:
import numpy as np 
import pandas as pd 

## Read in .dat file from logger and extract data 

In [2]:
# roof data 
file_r= '../files/taket_2015a_10.dat'
# bridge data 
file_b= '../files/bron_2014b_10.dat'

data = pd.read_table(file_r,  sep=',', header = 1, low_memory = False)
data_r = pd.read_table(file_r,  sep=',', header = 1, low_memory = False)
data_b = pd.read_table(file_b,  sep=',', header = 1, low_memory = False)


# skip first two rows 
data= data.iloc[2::,:]
# replace nan values with empty field
data = data.replace('NAN', '', regex=True)

In [4]:

data['TIMESTAMP'] = pd.to_datetime(data.TIMESTAMP)
#data['TIMESTAMP'] = data['TIMESTAMP'].dt.tz_localize('CET').dt.tz_convert('Europe/Stockholm')


In [10]:
data = data.iloc[0:1000,:]
import pytz
data.TIMESTAMP.dt.tz_localize(tz=pytz.FixedOffset(60)).dt.tz_convert('Europe/Stockholm')

2      2019-06-12 10:50:00+02:00
3      2019-06-12 11:00:00+02:00
4      2019-06-12 11:10:00+02:00
5      2019-06-12 11:20:00+02:00
6      2019-06-12 11:30:00+02:00
                  ...           
997    2019-12-26 09:50:00+01:00
998    2019-12-26 10:00:00+01:00
999    2019-12-26 10:10:00+01:00
1000   2019-12-26 10:20:00+01:00
1001   2019-12-26 10:30:00+01:00
Name: TIMESTAMP, Length: 1000, dtype: datetime64[ns, Europe/Stockholm]

In [12]:
data.RECORD.values

array(['0', '1', '2', ..., '2160', '2161', '2162'], dtype=object)

In [5]:
len(columns_bridge)

30

In [3]:
columns_roof = ['Wd_min_Avg', 'Wd_avg_Avg', 'Wd_max_Avg', 'Ws_min_Avg', 'Ws_avg_Avg',
       'Ws_max_Avg', 'Ta_Avg', 'RH_Avg', 'P_Avg','REGN_Tot','REGNDYGN',
       'SPN1_Tota_Avg', 'SPN1_diff_Avg', 'temp_L_K_Avg', 'L_sig_Avg',
      ]

columns_bridge = ['Wd_min_Avg', 'Wd_avg_Avg', 'Wd_max_Avg',
       'Ws_min_Avg', 'Ws_avg_Avg', 'Ws_max_Avg', 'Ta_Avg', 'RH_Avg', 'P_Avg','Rc_amount_Avg', 'Rd_dur_Avg', 'Ri_intens_Avg', 'Hc_amount_Avg',
       'Hd_dur_Avg', 'Hd_intens_Avg','Wd_min_Std', 'Wd_avg_Std', 'Wd_max_Std', 'Ws_min_Std', 'Ws_avg_Std',
       'Ws_max_Std', 'Ta_Std', 'RH_Std', 'P_Std', 'Rc_amount_Std',
       'Rd_dur_Std', 'Ri_intens_Std', 'Hc_amount_Std', 'Hd_dur_Std',
       'Hd_intens_Std']

meteodata_roof = ['wd_min', 'wd_avg', 'wd_max', 'ws_min', 'ws_avg', 'ws_max', 'T', 'RH', 'P', 'Precip', 'REGNDYGN',
       'SPN1_Tota_Avg', 'SPN1_diff_Avg', 'temp_L_K_Avg', 'L_sig_Avg']
meteodata_bridge = ['wd_min', 'wd_avg', 'wd_max', 'ws_min', 'ws_avg', 'ws_max', 'T', 'RH', 'P', 'Rc_amount_Avg', 'Rd_dur_Avg', 'Ri_intens_Avg', 'Hc_amount_Avg',
       'Hd_dur_Avg', 'Hd_intens_Avg','wd_max_std', 'wd_avg_std', 'wd_min_std', 'ws_min_std', 'ws_avg_std', 'ws_max_std', 'T_std', 'RH_Std', 'P_std', 'Rc_amount_std', 'Rd_dur_std', 'Ri_intens_std', 'Hc_amount_std',
       'Hd_dur_std', 'Hd_intens_std']

In [9]:
# save date as: Year, DOY, HHHMM 
data['time'] = pd.to_datetime(data.TIMESTAMP)
data['year'] = data.time.dt.year
data['hhmm']= data.time.dt.hour.astype(str) + data.time.dt.minute.astype(str)
data['doy'] = data.time.dt.dayofyear
data['code'] = data.RECORD.values

data= data[data.time.dt.year == 2020]

# create new data frame 
df= pd.DataFrame(dtype = 'object')
df['year'] = data.time.dt.year
df['hhmm']= data.time.dt.hour.astype(str) + data.time.dt.minute.astype(str)
df['doy'] = data.time.dt.dayofyear.astype(str)
df['code'] = data.RECORD.values.astype(str)

In [11]:
data.columns

Index(['TIMESTAMP', 'RECORD', 'LUFTTEMP_Avg', 'LUFTFUKT_Avg',
       'VINDHAST_WVc(1)', 'VINDHAST_WVc(2)', 'VINDHAST_WVc(3)', 'REGN_Tot',
       'REGNDYGN', 'Lufttryck_Avg', 'Ltryckhel_Avg', 'Ltrydec_Avg',
       'SPN1_Tota_Avg', 'SPN1_diff_Avg', 'temp_L_K_Avg', 'L_sig_Avg',
       'Wd_min_Avg', 'Wd_avg_Avg', 'Wd_max_Avg', 'Ws_min_Avg', 'Ws_avg_Avg',
       'Ws_max_Avg', 'Ta_Avg', 'RH_Avg', 'P_Avg', 'Rc_amount_Avg',
       'Rd_dur_Avg', 'Ri_intens_Avg', 'Hc_amount_Avg', 'Hd_dur_Avg',
       'Hd_intens_Avg', 'Ws_avg_WVc(1)', 'Ws_avg_WVc(2)', 'Ws_avg_WVc(3)',
       'BATTERI_Min', 'BATTERI_TMn', 'time', 'year', 'hhmm', 'doy', 'code'],
      dtype='object')

In [ ]:
## Meteo data Roof 

# relative humidity (%)
df['RH'] = data.RH_Avg.values.astype(float)
# pressure (hPa)
df['P'] = data.P_Avg.values.astype(float)
# mean air temperature (degC)
df['Ta']  = data.Ta_Avg.values
# mean wind speed (m/s)
df['ws']  = data.Ws_avg_Avg.values.astype(float)
# min wind speed (m/s)
df['ws_max']  = data.Ws_min_Avg.values.astype(float)
# max wind speed (m/s)
df['ws_min']  = data.Ws_max_Avg.values.astype(float)
# mean wind direction (deg)
df['wd']  = data.Ws_avg_Avg.values.astype(float)
# min wind direction (deg)
df['wd_max']  = data.Ws_min_Avg.values.astype(float)
# max wind direction (deg)
df['wd_min']  = data.Ws_max_Avg.values.astype(float)


In [13]:
## Meteo data Bridge 

# relative humidity (%)
df['RH'] = data.RH_Avg.values.astype(float)
# pressure (hPa)
df['P'] = data.P_Avg.values.astype(float)
# mean air temperature (degC)
df['Ta']  = data.Ta_Avg.values
# mean wind speed (m/s)
df['ws']  = data.Ws_avg_Avg.values.astype(float)
# min wind speed (m/s)
df['ws_max']  = data.Ws_min_Avg.values.astype(float)
# max wind speed (m/s)
df['ws_min']  = data.Ws_max_Avg.values.astype(float)
# mean wind direction (deg)
df['wd']  = data.Ws_avg_Avg.values.astype(float)
# min wind direction (deg)
df['wd_max']  = data.Ws_min_Avg.values.astype(float)
# max wind direction (deg)
df['wd_min']  = data.Ws_max_Avg.values.astype(float)

## standard deviations

# relative humidity (%)
df['RH_std'] = data.RH_Std.values.astype(float)
# pressure (hPa)
df['P_std'] = data.P_Std.values.astype(float)
# mean air temperature (degC)
df['Ta_std']  = data.Ta_Std.values
# mean wind speed (m/s)
df['ws_std']  = data.Ws_avg_Std.values.astype(float)
# min wind speed (m/s)
df['ws_max_std']  = data.Ws_min_Std.values.astype(float)
# max wind speed (m/s)
df['ws_min_std']  = data.Ws_max_Std.values.astype(float)
# mean wind direction (deg)
df['wd_std']  = data.Ws_avg_Std.values.astype(float)
# min wind direction (deg)
df['wd_max_std']  = data.Ws_min_Std.values.astype(float)
# max wind direction (deg)
df['wd_min_std']  = data.Ws_max_Std.values.astype(float)


## quality flags for meteorological variables 

# relative humidity (%)
df['RH_QC'] = 0
# pressure (hPa)
df['P_QC'] = 0
# mean air temperature (degC)
df['Ta_QC']  = 0
# mean wind speed (m/s)
df['ws_QC']  = 0
# min wind speed (m/s)
df['ws_max_QC']  = 0
# max wind speed (m/s)
df['ws_min_QC']  = 0
# mean wind direction (deg)
df['wd_QC']  = 0
# min wind direction (deg)
df['wd_max_QC']  = 0
# max wind direction (deg)
df['wd_min_QC']  = 0

## Quality control flags 

- 0: passed all quality tests 
- 1: not in plausible range
- 2: mutual exclusion/ inconsistency 
- 3: error in time evolution (not over minimum standard deviation of last hour) 
- 4: error in time consistency (abrupt jumps in values)




All quality tests are based on recommendations from the WMO (ref). 

In [267]:
# test with a small sample of dataframe 
subsample = df.head()
df= subsample

In [281]:
df['QC_flag'] = 1

for i in df.index.values:
    
    rh = df[df.index == i].RH.values[0]
    pr = df[df.index == i].P.values[0]
    ta = df[df.index == i].Ta.values[0]
    wd = df[df.index == i].wd.values[0]
    ws = df[df.index == i].ws.values[0]

    if i > 2 :
        p= i - 1 
        rh_p = df[df.index == p ].RH.values[0]
        pr_p = df[df.index == p ].P.values[0]
        ta_p = df[df.index == p].Ta.values[0]
        wd_p = df[df.index == p].wd.values[0]
        ws_p = df[df.index == p].ws.values[0]
        

    if i < np.shape(df.index.values)[0]:
        n= i+1
        rh_n = df[df.index == n].RH.values[0]
        pr_n = df[df.index == n].P.values[0]
        ta_n = df[df.index == n].Ta.values[0]
        wd_n = df[df.index == n].wd.values[0]
        ws_n = df[df.index == n].ws.values[0]

    # plausible ranges
    if rh < 0 or rh > 100:
         df.loc[i, 'QC_flag'] = 4
    if  pr < 500 or pr > 1100:
        df.loc[i, 'QC_flag'] = 4
    if wd < 0 or wd > 360:
        df.loc[i, 'QC_flag'] = 4      
    if ws < 0 or ws > 75:
        df.loc[i, 'QC_flag'] = 4
        
    # internal inconsistencies 
    if wd == 0 and ws > 0:
        df.loc[i, 'QC_flag'] = 4
    if ws == 0 and wd > 0:
        df.loc[i, 'QC_flag'] = 4
        
    # time consistency (check for jumps in data)
    ts= np.absolute(wd - wd_p) + np.absolute(wd - wd_n) 
    std= 4 * df[df.index == i].wd_std.values[0]
    if  ts > std :
        df.loc[i, 'QC_flag'] = 3

    ts= np.absolute(ws - ws_p) + np.absolute(ws - ws_n) 
    std= 4 * df[df.index == i].wd_std.values[0]
    if  ts > std :
        df.loc[i, 'QC_flag'] = 3

    ts= np.absolute(pr - pr_p) + np.absolute(pr - pr_n) 
    std= 4 * df[df.index == i].wd_std.values[0]
    if  ts > std :
        df.loc[i, 'QC_flag'] = 3

    ts= np.absolute(rh - rh_p) + np.absolute(rh - rh_n) 
    std= 4 * df[df.index == i].wd_std.values[0]
    if  ts > std :
        df.loc[i, 'QC_flag'] = 3
  
    # minimum standard deviation of last 60 minutes 
        


In [282]:
df.QC_flag

2    3
3    1
4    3
5    1
6    1
Name: QC_flag, dtype: int64

##  Saving format of monthly files 

In [ ]:
# check whether one month has passed 


In [ ]:
# create CSV file according to standard 

## Harmonizing historical data 

- same columns in all monthly files 
- flags for quality control 
- metadata  (location, explanation of variable names, units) in a header at the top  of the file with # to indicate these are comments. A blank line with  leading # to separate the metadata from the column headers
- make  the actual column headers legal variable names eg T_max, not “Maximum  temperature (degC)”.
- Put the explanation in the header.
-   column header line also has a leading 
- use a ; separator, dont use fixed-width or space separated. 
-  indicate  missing values by not having a value! eg ;; Now, test that R and pandas  are ok with this, but it will reduce the file-size dramatically  compared with -999 or “NaN”. Currently the files in BridgeData total about 600MB, which is absurd. 
- similarly, don’t use more decimal places than the precision of the actual measurements.
- UTF-8 with linefeed

## write header 

In [140]:
headerstr= 'Roof station GVC \nLon: 11.966434 \nLat: 57.688374 \nTa: Air temperature [degC], RH: Relative Humidity [%], P: Air pressure [hpa], Ws: 10min mean wind speed [m/s], Wind direction [deg], \nQC: quality flags (0 = passed all controls, 1 = not in plausible range, 2 = inconsistency, 3 = same value, 4 = big jump)'


In [174]:
print(headerstr)

Roof station GVC 
Lon: 11.966434 
Lat: 57.688374 
Ta: Air temperature [degC], RH: Relative Humidity [%], P: Air pressure [hpa], Ws: 10min mean wind speed [m/s], Wind direction [deg], 
QC: quality flags (0 = passed all controls, 1 = not in plausible range, 2 = inconsistency, 3 = same value, 4 = big jump)


**Notes Julia - Identify usable columns in historical data files: ** 

- Ta (Ta.1?)
- RH (RH.1?)
- P
- Wd_avg, Wd_min, Wd_max , Wd_std
- Ws_avg, Ws_min, Ws_max
- radiation columns?? 